### PR TITLE
Entities reference guide

### LINKED ISSUE #198: [Reference] Entities
Create an "Entities" section in Sensu 2.0 "Reference": #149

### PR DESCRIPTION
**Pull request by nikkiattea**
Signed-off-by: Nikki Attea <nikki@sensu.io>

Reference guide documentation for Entities.
Closes #198

### CODE REVIEW

File: content/sensu-core/2.0/reference/entities.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+# Entities
+
+### Why "Entities"?
+
+Sensu 2.0 uses an updated data model to address the latest paradigm shift in infrastructure and software delivery like containers and "serverless". Sensu is no longer confined to the concept of a client, we have "entities" that can can represent anything.
+
+### What is a Sensu entity?
+
+Entities, formally known as clients in Sensu 1.x, are monitoring agents (e.g. server, container, etc) that need to be monitored. They contain system information such as the hostname, OS, platform, and version. Entities not only provide context to event data but an event's uniqueness is determined by the check name and the ID of the entity upon which the check ran.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: The first sentence is hard to parse for me. I think I don't get "monitoring agents (e.g. server, container, etc) that need to be monitored." Maybe removing the server/container detail or mentioning whether each entity is a device monitored or it implies a running agent.

File: content/sensu-core/2.0/reference/entities.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+## Agent entities
+
+### What is an agent entity?
+
+An agent entity is a Sensu entity which falls under the "agent" class. Agent entities are monitoring agents, which are installed and run on every system that needs to be monitored. The entity is responsible for registering the system with Sensu, sending keepalive messages (the Sensu heartbeat mechanism), and executing monitoring checks. Each entity is a member of one or more subscriptions – a list of roles and/or responsibilities assigned to the system (e.g. a webserver, database, etc). Sensu entities will "subscribe" to (or watch for) check requests published by the Sensu server (via the Sensu Transport), execute the corresponding requests locally, and publish the results of the check back to the transport (to be processed by a Sensu server).
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: First sentence ("An agent entity is a Sensu entity which falls under the "agent" class") is programming info that doesn't make it clearer to a user. I think it can be removed unless you want to reword to in a helpful way to the user.

File: content/sensu-core/2.0/reference/entities.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+### Creating an agent entity
+
+To create an agent entity, Sensu 2 must be installed. Upon spinning up the sensu-agent, you can use several flags to customize the properties of the entity. In the example below, the optional flags are used to describe the default values of the agent entity that gets created upon starting sensu-agent. Other flags/properties you can set for your agent entity include: --deregister, --subscriptions, --deregistration-handler, and --redact.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: On versioning: Sensu 2.0 or greater? Might be a larger question we can deal with later 😄

File: content/sensu-core/2.0/reference/entities.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+{{< highlight shell >}}
+sensu-agent start --id hostname --environment default --organization default --keepalive-timeout 120 --user agent
+{{< /highlight >}}
+
+## Proxy Entities
+
+### What is a proxy entity?
+
+A proxy entity is a Sensu entity which falls under the "proxy" class. Proxy entities (formerly known as proxy clients, "Just-in-time" or "JIT" clients) are dynamically created entities, added to the entity registry if an entity does not already exist for a check result. Proxy entity registration differs from keepalive-based registration because the registration event happens while processing a check result (not a keepalive message). Sensu proxy entities allow Sensu to monitor external resources (e.g. on systems and/or devices where a sensu-agent cannot be installed, such a network switches), using the defined check ProxyEntityID to create a proxy entity for the external resource. Once created, proxy entities work much in the same way as any other Sensu entity.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: class statement again - I think there's enough info while deleting the first sentence.

File: content/sensu-core/2.0/reference/entities.md (hunk 1 above)
Comment (mbbroberg) [reply content/sensu-core/2.0/reference/entities.md]: I also think speaking in the singular under this section would be nice to clarify (vs plural everywhere else).

File: content/sensu-core/2.0/reference/entities.md (hunk 2 above)
Comment (nikkiattea) [reply content/sensu-core/2.0/reference/entities.md]: `Class` is actually a field of an Entity that is visible to the user, which describes whether it is an "agent" entity or "proxy" entity. But let me reword that so the point is clear, I totally see where you're coming from with the use of "class."

Comment (grepory) [reply content/sensu-core/2.0/reference/entities.md]: FWIW, users can specify classes of entities that are not "agent" or "proxy." Those are just reserved words for us to use internally.

Comment (nikkiattea) [reply content/sensu-core/2.0/reference/entities.md]: So it sounds like the contents of that field are trivial in this section then ;)

File: content/sensu-core/2.0/reference/entities.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+menu:
+  sensu-core-2.0:
+    parent: reference
+---
+
+# Entities
+
+### Why "Entities"?
+
+Sensu 2.0 uses an updated data model to address the latest paradigm shift in infrastructure and software delivery like containers and "serverless". Sensu is no longer confined to the concept of a client, we have "entities" that can can represent anything.
```
[the remark this answers is not in the document]
Comment (nikkiattea) [reply content/sensu-core/2.0/reference/entities.md]: @mbbroberg Thoughts?

> Sensu 2.0 uses an updated data model to address the latest paradigm shift in infrastructure and software delivery like containers and "serverless", so we've adjusted the nomenclature accordingly. Sensu is no longer confined to the concept of a client, we have "entities" that can can represent anything.

File: content/sensu-core/2.0/reference/entities.md [hunk 6]
```diff
@@ -14,21 +14,21 @@ menu:
 
 ### Why "Entities"?
 
-Sensu 2.0 uses an updated data model to address the latest paradigm shift in infrastructure and software delivery like containers and "serverless". Sensu is no longer confined to the concept of a client, we have "entities" that can can represent anything.
+Sensu 2.0 uses an updated data model to address the latest paradigm shift in infrastructure and software delivery like containers and "serverless", so we've adjusted the nomenclature accordingly. Sensu is no longer confined to the concept of a client, we have "entities" that can can represent anything.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: I feel like we can get even more direct by acknowledging the need for new "thing" to be monitored. Maybe: 

Sensu 2.0 uses an updated data model that allows for it to acknowledge the full range of infrastructure, runtime and application types that compose a complete monitoring environment (from server hardware to "serverless" functions). Sensu no longer focuses on the object doing the monitoring (previously called a "client") and instead focuses on the object it monitors. We call these monitored parts of an infrastructure "entities."

File: content/sensu-core/2.0/reference/entities.md [hunk 7]
```diff
@@ -14,21 +14,21 @@ menu:
...
 
 ### What is a Sensu entity?
 
-Entities, formally known as clients in Sensu 1.x, are monitoring agents (e.g. server, container, etc) that need to be monitored. They contain system information such as the hostname, OS, platform, and version. Entities not only provide context to event data but an event's uniqueness is determined by the check name and the ID of the entity upon which the check ran.
+An Entity, formally known as a client in Sensu 1.x, represents anything (e.g. server, container, network switch, etc) that needs to be monitored. It can contain system information such as the hostname, OS, platform, and version. An Entity not only provides context to event data (what/where the event is from) but an event's uniqueness is determined by the check name and the ID of the entity upon which the check ran.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: I like this a lot, especially if you adopt some form of the comment above that leaves the user with the concept of entities. 

Minor things: we will want to consistently capitalize keywords (Entity, Entities) or not. Right now we have a mix. Based on what I see so far, I'd recommend sticking with lowercase and maybe **bolding** new concepts when you introduce them (instead of capitalizing them).

File: content/sensu-core/2.0/reference/entities.md [hunk 8]
```diff
@@ -14,21 +14,21 @@ menu:
...
+Agent entities are monitoring agents, which are installed and run on every system that needs to be monitored. The entity is responsible for registering the system with Sensu, sending keepalive messages (the Sensu heartbeat mechanism), and executing monitoring checks. Each entity is a member of one or more subscriptions – a list of roles and/or responsibilities assigned to the system (e.g. a webserver, database, etc). Sensu entities will "subscribe" to (or watch for) check requests published by the Sensu server (via the Sensu Transport), execute the corresponding requests locally, and publish the results of the check back to the transport (to be processed by a Sensu server).
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: Since we're being super clear here, maybe change "registering the system with Sensu" to acknowledge the specific service it registers with, such as "registering the system with Sensu backend service" if that's accurate.  (I need your help there 😄). 

Love the additional explanation of the heartbeat mechanism - great details in here 👍

Comment (mbbroberg) [reply content/sensu-core/2.0/reference/entities.md]: Maybe also: 

- bold **subscriptions** since that's a new concept we want to highlight

Comment (mbbroberg) [reply content/sensu-core/2.0/reference/entities.md]: > subscriptions – a list of roles and/or responsibilities assigned to the system

Are these assigned to the system or, more specifically, to the agent entity? I'm not sure what "the system" means in this context.

File: content/sensu-core/2.0/reference/entities.md [hunk 9]
```diff
@@ -14,21 +14,21 @@ menu:
...
 
 ### Creating an agent entity
 
-To create an agent entity, Sensu 2 must be installed. Upon spinning up the sensu-agent, you can use several flags to customize the properties of the entity. In the example below, the optional flags are used to describe the default values of the agent entity that gets created upon starting sensu-agent. Other flags/properties you can set for your agent entity include: --deregister, --subscriptions, --deregistration-handler, and --redact.
+To create an agent entity, Sensu 2.x must be installed. Upon spinning up the sensu-agent, you can use several flags to customize the properties of the entity. In the example below, the optional flags are used to describe the default values of the agent entity that gets created upon starting sensu-agent. Other flags/properties you can set for your agent entity include: --deregister, --subscriptions, --deregistration-handler, and --redact.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: When talking about the agent, it's best to refer to it as the agent or mention the package `sensu-agent` with syntax highlighting. Bonus points for highlighting the flags as well. 

Side note - will these be consistent enough that you want to list them or can we link to where they will be kept up to date?

File: content/sensu-core/2.0/reference/entities.md [hunk 10]
```diff
@@ -38,7 +38,7 @@ sensu-agent start --id hostname --environment default --organization default --k
...
+Proxy entities (formerly known as proxy clients, "Just-in-time" or "JIT" clients) are dynamically created entities, added to the entity registry if an entity does not already exist for a check result. Proxy entity registration differs from keepalive-based registration because the registration event happens while processing a check result (not a keepalive message). Sensu proxy entities allow Sensu to monitor external resources (e.g. on systems and/or devices where a sensu-agent cannot be installed, such a network switches), using the defined check ProxyEntityID to create a proxy entity for the external resource. Once created, proxy entities work much in the same way as any other Sensu entity.
```
Comment (mbbroberg) [inline content/sensu-core/2.0/reference/entities.md]: First time I'm hearing of the registry. Is that something that could be intro'ed in "What is a Sensu entity?" Maybe something like: "Entities are stored in a registry managed by xyz" 

Could be the last sentence of that entry.

**Comment by palourde** [comment]
LGTM
